### PR TITLE
Segments and softmaxes.

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
- Support `--n-segments` in polyencoder
- Use fp16 safe cross entropy loss for the ranking objective.

**Testing steps**
Internal sweeps.

### CODE REVIEW

**Comment by stephenroller** [comment]
Ya I yolo'd the fp16 change, I'm still testing it internally.

**Comment by stephenroller** [comment]
It worked well internally.
